util/log: TestSecondaryGC failed

[(util/log).TestSecondaryGC failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2491255&tab=buildLog) on [master@5d5a8e2f11db706f48d4b0d5489967875c98c3fb](https://github.com/cockroachdb/cockroach/commits/5d5a8e2f11db706f48d4b0d5489967875c98c3fb):

```
=== RUN   TestSecondaryGC
    file_log_gc_test.go:39: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSecondaryGC324526730
    file_log_gc_test.go:156: expected 20 files, but found 19
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestSecondaryGC324526730
--- FAIL: TestSecondaryGC (0.07s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestSecondaryGC PKG=./pkg/util/log TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestSecondaryGC.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/util/log/file_log_gc.go
+++ b/pkg/util/log/file_log_gc.go
@@ -21,7 +21,6 @@ import (
 
 // gcDaemon runs the GC loop for the given logger.
 func (l *fileSink) gcDaemon(ctx context.Context) {
-	l.gcOldFiles()
 	for {
 		select {
 		case <-ctx.Done():

--- a/pkg/util/log/file_log_gc_test.go
+++ b/pkg/util/log/file_log_gc_test.go
@@ -88,8 +88,6 @@ func testLogGC(
 	}(logging.mu.disableDaemons)
 	logging.mu.Unlock()
 
-	const newLogFiles = 20
-
 	// Make an entry in the target logger. This ensures That there is at
 	// least one file in the target directory for the logger being
 	// tested. This serves two
@@ -132,6 +130,8 @@ func testLogGC(
 	// Pick a max total size that's between 2 and 3 log files in size.
 	maxTotalLogFileSize := logFileSize*expectedFilesAfterGC + logFileSize // 2
 
+	t.Logf("new max total log file size: %d", maxTotalLogFileSize)
+
 	// We want to create multiple log files below. For this we need to
 	// override the size/number limits first to the values suitable for
 	// the test.
@@ -143,6 +143,8 @@ func testLogGC(
 	atomic.StoreInt64(&fileSink.logFilesCombinedMaxSize, maxTotalLogFileSize)
 
 	// Create the number of expected log files.
+	const newLogFiles = 20
+
 	for i := 1; i < newLogFiles; i++ {
 		logFn(context.Background(), "%d", i)
 		Flush()

--- a/pkg/util/log/flags.go
+++ b/pkg/util/log/flags.go
@@ -179,8 +179,10 @@ func ApplyConfig(config logconfig.Config) (cleanupFn func(), err error) {
 		allSinkInfos.put(fileSinkInfo)
 
 		if fileSink.logFilesCombinedMaxSize > 0 {
+			// Do a start round of GC, so clear up past accumulated files.
+			fileSink.gcOldFiles()
 			// Start the GC process. This ensures that old capture files get
-			// erased as necessary.
+			// erased as new files get created.
 			go fileSink.gcDaemon(secLoggersCtx)
 		}
 

--- a/pkg/util/log/flags.go
+++ b/pkg/util/log/flags.go
@@ -225,7 +225,7 @@ func ApplyConfig(config logconfig.Config) (cleanupFn func(), err error) {
 	}
 
 	// Apply the stderr sink configuration.
-	logging.stderrSink.noColor = config.Sinks.Stderr.NoColor
+	logging.stderrSink.noColor.Set(config.Sinks.Stderr.NoColor)
 	if err := logging.stderrSinkInfoTemplate.applyConfig(config.Sinks.Stderr.CommonSinkConfig); err != nil {
 		cleanupFn()
 		return nil, err
@@ -356,7 +356,7 @@ func DescribeAppliedConfig() string {
 	}
 
 	// Describe the stderr sink.
-	config.Sinks.Stderr.NoColor = logging.stderrSink.noColor
+	config.Sinks.Stderr.NoColor = logging.stderrSink.noColor.Get()
 	config.Sinks.Stderr.CommonSinkConfig = logging.stderrSinkInfoTemplate.describeAppliedConfig()
 
 	describeConnections := func(l *loggerT, ch Channel,

--- a/pkg/util/log/format_crdb_v1.go
+++ b/pkg/util/log/format_crdb_v1.go
@@ -62,7 +62,7 @@ func (formatCrdbV1TTY) formatterName() string { return "crdb-v1-tty" }
 
 func (formatCrdbV1TTY) formatEntry(entry logpb.Entry, stacks []byte) *buffer {
 	cp := ttycolor.StderrProfile
-	if logging.stderrSink.noColor {
+	if logging.stderrSink.noColor.Get() {
 		cp = nil
 	}
 	return formatLogEntryInternal(entry, false /*showCounter*/, cp, stacks)
@@ -77,7 +77,7 @@ func (formatCrdbV1TTYWithCounter) formatterName() string { return "crdb-v1-tty-c
 
 func (formatCrdbV1TTYWithCounter) formatEntry(entry logpb.Entry, stacks []byte) *buffer {
 	cp := ttycolor.StderrProfile
-	if logging.stderrSink.noColor {
+	if logging.stderrSink.noColor.Get() {
 		cp = nil
 	}
 	return formatLogEntryInternal(entry, true /*showCounter*/, cp, stacks)

--- a/pkg/util/log/stderr_sink.go
+++ b/pkg/util/log/stderr_sink.go
@@ -10,13 +10,16 @@
 
 package log
 
-import "github.com/cockroachdb/cockroach/pkg/cli/exit"
+import (
+	"github.com/cockroachdb/cockroach/pkg/cli/exit"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+)
 
 // Type of a stderr copy sink.
 type stderrSink struct {
 	// the --no-color flag. When set it disables escapes code on the
 	// stderr copy.
-	noColor bool
+	noColor syncutil.AtomicBool
 }
 
 // activeAtSeverity implements the logSink interface.

